### PR TITLE
Improve docs for `from_dlpack` and `to_dlpack`

### PR DESCRIPTION
**Pull request by rgommers**
This moves the warning to the legacy function where it belongs, improves the phrasing, and adds examples.

There may be more to do to make `from_dlpack` more discoverable as a follow-up, because in multiple issues/PR we discovered people wanted new things (e.g., a memoryview-like object, or `__array_interface__` support) that `from_dlpack` already provides.

### CODE REVIEW

**Comment by rgommers** [comment]
Hmm, the docs preview job has been removed and https://hud.pytorch.org/pr/70437 wants so many permissions that I can't access it. This should be good to go, but I can't verify (local KaTeX has a problem for me).

**Comment by vadimkantorov** [comment]
It mentions "old-style DLPack capsule", but is there a "new-style DLPack capsule"? I think it'd be a frequent question

**Comment by rgommers** [comment]
> It mentions "old-style DLPack capsule", but is there a "new-style DLPack capsule"? I think it'd be a frequent question

the "new style DLPack" is to just pass a tensor directly. It's still the same capsule, it's now just under the hood, so better UX and no change of consuming it twice.

Could you propose a rephrase?

**Comment by vadimkantorov** [comment]
Well, I still don't understand the meaning of old/new style :( We often need a DLPack tensor to develop an interoperable C interface (not only for exchanging tensors between NumPy / Deep Learning frameworks), like I did here: https://github.com/vadimkantorov/dotnetdlpack

From phrasing `to_dplack returns a new-style capsule` is still opaque to me in terms of what's a capsule or what's inside or how it's what is new/old style.

**Comment by rgommers** [comment]
To explain:
- There is only one kind of "capsule". DLPack has been stable for years, no change in format there.
- At the Python level this was the old usage pattern `capsule = to_dlpack(t)` followed by `t_new = from_dlpack(capsule)`. There is nothing else you can do with `capsule` other than pass it to `from_dlpack` _exactly_ once.
- New style is `t_new = from_dlpack(t)`. the exact same "capsule" is produced under the hood (by `t.__dlpack__`), used and then discarded safely. This UX is strictly better, and nothing else changes.

> is still opaque to me in terms of what's a capsule or what's inside

It should be, because it's pretty much useless. The name is derived from `PyCapsule`, which is a Python C API thing that a user shouldn't have to know about. When the versions of both libraries you are using support `__dlpack__` (and hence `t_new = from_dlpack(t)`), then you can forget all about `capsule`. If you have older versions, you will still need it.

**Comment by vadimkantorov** [comment]
I think I understand now, but it should be explicitly explained what is the new and old style, or remove the refernece to this concepts at all and just explain that two types of arguments are accepted.

In my case, I needed DLPack not for interop between frameworks but to create a .NET wrapper consuming DLPack API using libtorch. So I needed to understand exactly what is the capsule and what is the field name. All this is better be explained. There was an issue in DLPack complaining that the name of the field inside the capsule is not documented: https://github.com/dmlc/dlpack/issues/51. It would be better if PyTorch documented it as well or referenced some ground-truth documentation. It may also be nice to include a snippet of manual working with these capsules. This example is very useful if you are not a framework developer. E.g. I had to figure this out in https://github.com/vadimkantorov/pydlpack and figuring out the proper semantics of DLTensor deallocator and its interaction with capsule is quite non-trivial.

**Comment by rgommers** [comment]
> It would be better if PyTorch documented it as well or referenced some ground-truth documentation.

That belongs in the DLPack docs, it doesn't make sense to explain the kind of details you are mentioning here in the PyTorch Python API docs - they are irrelevant to Python users. You are only running into it because you are using C/C++ APIs.

The minor catch is that DLPack currently has no docs (outside of inline comments). I plan to address that in H1 2022. For now this is probably the best user-facing DLPack documentation: https://data-apis.org/array-api/latest/design_topics/data_interchange.html#dlpack-support. Most of that should move to DLPack itself in the future.

> what is the new and old style, or remove the refernece to these concepts at all

Then I'd prefer to just tweak the text a bit to remove "old style", and only state that using `from_dlpack(tensor)` is the preferred usage.

**Comment by vadimkantorov** [comment]
> `from_dlpack(tensor)` is the preferred usage.

Why not! Just please specify that the tensor object should have a `__dlpack__` capsule property. So that the "old-style" usages where we manually prepare the PyCapsule can figured out.

**Comment by rgommers** [comment]
> A bit disagreeing. I'm running into it very much because DLPack is for doing interop with PyTorch's Python API, and interop with custom C code is a big part of it.

Okay, interesting. Not quite agreeing that we should include content that is irrelevant for Python users and only needed if you write custom C code, but I understand better what you're doing at least after I had a look at your repo.

Side note: NumPy 1.22.0 will have support for DLPack, so your custom code for that will become obsolete - it'll just be `x = np.from_dlpack(tensor)`.

**Comment by vadimkantorov** [comment]
Yeah, I also needed this when I was doing a "framework-agnostic" pure C module for reading audio with ffmpeg: https://github.com/vadimkantorov/readaudio

File: torch/utils/dlpack.py [hunk 1]
```diff
@@ -22,33 +22,78 @@ class DLDeviceType(enum.IntEnum):
...
+  ``to_dlpack`` is a legacy DLPack interface. The capsule it returns
+  cannot be used for anything in Python other than use it as input to
+  ``from_dlpack``. The more idiomatic use of DLPack is to call
+  ``from_dlpack`` directly on the tensor object - this works when that
+  object has a ``__dlpack__`` method, which PyTorch and most other
+  libraries indeed have now.
+
+.. warning::
+  Only call ``from_dlpack`` once per capsule produced with ``to_dlpack``.
+  Behavior when a capsule is consumed multiple times is undefined.
```
Comment (leofang) [inline torch/utils/dlpack.py:37]: Nitpick: technically it's well defined (an error would be raised if the capsule has been consumed with its name changed) long before we settled on `__dlpack__`, but I'm fine leaving it as is.

**Review by leofang — APPROVED** [review]
LGTM!

**Review by mruberry — APPROVED** [review]
Sorry to keep you waiting, @rgommers. Let's try to get this merged!

**Pull request closed without merging**